homeassistant/components/cover/__init__.py

```diff
@@ -484,12 +484,12 @@ async def async_toggle_tilt(self, **kwargs: Any) -> None:
     def _get_toggle_function(
         self, fns: dict[str, Callable[_P, _R]]
     ) -> Callable[_P, _R]:
-        if CoverEntityFeature.STOP | self.supported_features and (
+        if CoverEntityFeature.STOP in self.supported_features and (
             self.is_closing or self.is_opening
         ):
             return fns["stop"]
-        if self.is_closed:
+        if self.is_closed or self.is_closing:
             return fns["open"]
-        if self._cover_is_last_toggle_direction_open:
+        if self.is_opening or self._cover_is_last_toggle_direction_open:
```

Comment:
`is_opening` is supposed to always match `_cover_is_last_toggle_direction_open` after a state update. That's why we didn't consider `is_opening` and `is_closing` in these two checks.

I think we should remove `_cover_is_last_toggle_direction_open` and use `is_opening` and `is_closing` instead so we avoid a possible race condition or state inconsistency.

Reply:
I don't agree, because not all covers report `is_opening` and `is_closing`.

I would suggest to split this PR in two:
1. A PR which fixes the incorrect check for `CoverEntityFeature.STOP` and adds tests for that
2. A PR which extends or changes the logic to decide if the next direction should be open or close

Reply:
`_cover_is_last_toggle_direction_open` isn't used anywhere but here and not documented. What are you suggesting?

Reply:
I think I agree with @emontnemery , in that we cannot just remove `_cover_is_last_toggle_direction_open` without some additional tweaks. That would cause us to _only_ select the "close" function if `is_opening`. And that seems clearly wrong, such as when the state is fully open.

I think it also serves an important function in that it "remembers" whether we were most recently opening or closing, even after we have stopped. Without it, when stopped partially open, `toggle()` can only ever send it in the same direction, every time. With it, we support alternating sequences like:
* Initial state is open
* `toggle()` -> `close_cover()`; state becomes closing
* `toggle()` ->`stop_cover()` state becomes open again (partially)
* `toggle()` -> `open_cover()`; state becomes opening
* `toggle()` ->`stop_cover()` state becomes open again (partially)
* `toggle()` -> `close_cover()`; state becomes closing

I can split the changes up, if that is what we want. I decided to include everything because fixing the first (`STOP`) condition is going to start sending execution down these paths, that previously did not. I did not want to end up replacing an incorrect no-op behavior with an even more incorrect behavior, so making sure these later conditions are correct for the new cases that will now reach them seems an important part of the fix.

Whichever PR it goes in, does this logic seem better/clearer?
```python
# If we are opening or closing and we support stopping, then we should stop
if CoverEntityFeature.STOP in self.supported_features and (
    self.is_closing or self.is_opening
):
    return fns["stop"]

# If we are fully closed or in the process of closing, then we should open
if self.is_closed or self.is_closing:
    return fns["open"]

# If we are fully open or in the process of opening, then we should close
if self.current_cover_position == 100 or self.is_opening:
    return fns["close"]

# We are any of:
# * fully open but do not report `current_cover_position`
# * stopped partially open
# * either opening or closing, but do not report them
# If we previously reported opening/closing, we should move in the opposite direction.
# Otherwise, we must assume we are (partially) open and should always close.
# Note: _cover_is_last_toggle_direction_open will always remain True if we never report opening/closing.
return fns["close"] if self._cover_is_last_toggle_direction_open else fns["open"]
```

I think all three versions are actually logically equivalent, under the assumption that `state` is called often enough to keep `_cover_is_last_toggle_direction_open` updated. My intent in explicitly checking for opening/closing was to avoid the potential race situation where it could be stale. The version in this comment also better handles some possible weird state transitions, such as going from "closing" directly to "100% open". (The previous versions would try to open in that case.)

Reply:
I missed the remembered state feature of the attribute `_cover_is_last_toggle_direction_open`, which is important both if the cover stops somewhere between open and closed and when stopping at fully open.

I think it's ok to split the PR or keep it here. @emontnemery can decide.

I like the last version in the comment above since it has the same kind of checks for both directions and for both endpoints, closed and fully open. Good to add comments too, since it obviously is hard to keep track of all the cases.

Reply:
Let's please split it. The bug using incorrect logic to check for `CoverEntityFeature.STOP` should be fixed and included in a 2024.1.x patch release, the other changes could go in 2024.2 if accepted.

Reply:
I also think the last version with very detailed comments explaining the logic looks great 👍

Reply:
Okay, split things out and rebased on the latest `dev`. This PR now has only the minimal `STOP` fix and the tests for verifying it.

I pushed the remaining cleanup/fixes to [this branch](https://github.com/home-assistant/core/compare/dev...vexofp:homeassistant-core:106791-cover-toggle-closing-open) (which builds off of this commit to avoid conflicts). I will make the PR for that after this one merges.